### PR TITLE
Cvxpy

### PR DESCRIPTION
**Pull request by teh**
###### Motivation for this change

Add cvxpy + dependencies

Finishing off the cvxpy packaging mentioned in https://github.com/NixOS/nixpkgs/pull/45047

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/python-modules/cvxpy/default.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    inherit pname version;
+    sha256 = "11hdrgdcb7idncrbr00rmrxhi7f8sjm134x4wgqkkxwppklk36xv";
+  };
+
+  # Tests not discovered correctly on python3 unless nosetests invoked
+  # explicitly.
+  checkPhase = "nosetests -v";
+
+  meta = {
+    description = "A Python-embedded modeling language for convex optimization problems.";
```
Comment (costrouc) [inline pkgs/development/python-modules/cvxpy/default.nix]: Remove period from description

File: pkgs/development/python-modules/cvxpy/default.nix [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+, scs
+, multiprocess
+, fastcache
+, six
+, numpy
+, scipy
+, nose
+, lib
+}:
+buildPythonPackage rec {
```
Comment (costrouc) [inline pkgs/development/python-modules/cvxpy/default.nix]: newline

File: pkgs/development/python-modules/ecos/default.nix [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  ];
+  checkInputs = [ nose ];
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0n67plhclbsxqi23is8rj62vv3d75nnwzhsmya9jlbpknd10zsc3";
+  };
+
+  meta = {
+    description = "This is the Python package for ECOS: Embedded Cone Solver.";
```
Comment (costrouc) [inline pkgs/development/python-modules/ecos/default.nix]: remove period from description

**Review by costrouc — CHANGES_REQUESTED** [review]
Small fixes required but otherwise good. `nix-review` passed

File: pkgs/development/python-modules/osqp/default.nix [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+  nativeBuildInputs = [ cmake ];
+
+  propagatedBuildInputs = [ numpy future scipy ];
+
+  src = fetchFromGitHub {
+    owner = "oxfordcontrol";
+    repo = "osqp-python";
+    rev = "v0.5.0";
+    fetchSubmodules = true;
```
Comment (FRidh) [inline pkgs/development/python-modules/osqp/default.nix]: Do we really need this? Can we not package that one separately?

Comment (teh) [reply pkgs/development/python-modules/osqp/default.nix]: I could download the source separately but AFAICT the python module needs the osqp source, i.e. it's not just linking against a library [1].

https://github.com/oxfordcontrol/osqp-python/blob/master/setup.py#L95

File: pkgs/development/python-modules/ecos/default.nix (hunk 3 above)
Comment (teh) [reply pkgs/development/python-modules/ecos/default.nix]: done

File: pkgs/development/python-modules/cvxpy/default.nix (hunk 2 above)
Comment (teh) [reply pkgs/development/python-modules/cvxpy/default.nix]: done

File: pkgs/development/python-modules/cvxpy/default.nix (hunk 1 above)
Comment (teh) [reply pkgs/development/python-modules/cvxpy/default.nix]: done

**Review by teh — COMMENTED** [review]
thanks for the review!

**Comment by stites** [comment]
Heads up, this PR can fail at test time, possibly due to bad seeds:

```
======================================================================
FAIL: test_ch8_mcsdp (test_examples.TestExamples)
----------------------------------------------------------------------
Traceback (most recent call last):                     
  File "/build/cvxopt-1.2.3/tests/test_examples.py", line 48, in test_ch8_mcsdp
    self.assertAlmostEqualLists(list(z[::n+1]),n*[1.0],5)
  File "/build/cvxopt-1.2.3/tests/test_examples.py", line 19, in assertAlmostEqualLists
    for u,v in zip(L1,L2): self.assertAlmostEqual(u,v,places)
AssertionError: 3.1962078445048556 != 1.0 within 5 places
----------------------------------------------------------------------
Ran 39 tests in 172.236s
FAILED (failures=1, skipped=6)
```

This is the second time I've run this build. The first time the value was ~1.2

**Comment by teh** [comment]
That's pretty bad! `cvxopt` isn't part of this branch, it's been in master for a while. Not sure why it got triggered as build-from-source for you. I tried this (as a trusted user, e.g. root):

```
nix-build -A python3Packages.cvxopt --option repeat 10 --option substitute false  --no-out-link
these derivations will be built:
  /nix/store/wh6ikia2ck8xqwl2ljnifibrldsbpbvy-cvxopt-1.2.3.tar.gz.drv
  /nix/store/x9c5sc969yfwg2yjzi01m8w8x6dp2hrf-python3.7-cvxopt-1.2.3.drv
building '/nix/store/wh6ikia2ck8xqwl2ljnifibrldsbpbvy-cvxopt-1.2.3.tar.gz.drv'...
...
building '/nix/store/x9c5sc969yfwg2yjzi01m8w8x6dp2hrf-python3.7-cvxopt-1.2.3.drv' (round 1/11)...
...
building '/nix/store/x9c5sc969yfwg2yjzi01m8w8x6dp2hrf-python3.7-cvxopt-1.2.3.drv' (round 11/11)...
...
OK (skipped=6)
/nix/store/47x71dgcpf7h9214wr8hkjmz1pma3sk7-python3.7-cvxopt-1.2.3
```

And can't reproduce. Can you share your build command?

**Comment by stites** [comment]
Oops! cvxopt works fine on unstable, actually. The problem was that I was pulling cvxopt into a shell from this PR (which is broken) and I was too rushed to tell the differentiate between the suffixes. Unfortunately, I am still getting a test error with cvxpy : /


```
======================================================================
ERROR: test_psd_nsd_parameters (cvxpy.tests.test_expressions.TestExpressions)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/cvxpy-1.0.21/cvxpy/tests/test_expressions.py", line 265, in test_psd_nsd_parameters
    p.value = a2
  File "/build/cvxpy-1.0.21/cvxpy/expressions/constants/parameter.py", line 60, in value
    self._value = self._validate_value(val)
  File "/build/cvxpy-1.0.21/cvxpy/expressions/leaf.py", line 428, in _validate_value
    "%s value must be %s." % (self.__class__.__name__, attr_str)
ValueError: Parameter value must be positive semidefinite.

----------------------------------------------------------------------
Ran 1882 tests in 12.856s

FAILED (SKIP=2, errors=1)
Error in AA type 1, iter: 39, info: 0, norm 4.22e+04
Error in AA type 1, iter: 39, info: 0, norm 4.22e+04
Error in AA type 1, iter: 10, info: 0, norm 1.24e+04
builder for '/nix/store/i9y43yq923acwvxl3xamp1pczagw9mc4-python3.6-cvxpy-1.0.21.drv' failed with exit code 1                                
cannot build derivation '/nix/store/mlbl4p4a5vy99dans7zkhsiw85j6szwj-python3-3.6.8-env.drv': 1 dependencies couldn't be built
error: build of '/nix/store/mlbl4p4a5vy99dans7zkhsiw85j6szwj-python3-3.6.8-env.drv' failed                                                  
```

I pinned my nix-shell to your PR. Here's a compressed shell file:

```nix
# shell.nix
{ ... }:

let
  hostpkgs = import <nixpkgs> {};
  srcDef = builtins.fromJSON ''{
    "url": "https://github.com/NixOS/nixpkgs",
    "rev": "0115285de38b27ed9c3cf900ec3abff0a7a37632",
    "date": "2019-04-19T11:38:25+01:00",
    "sha256": "05yb5ww7wl68v4fwq0fg6k6sdf49ykfr7yckhlyx2a1glzpdzs1h",
    "fetchSubmodules": false
  }'';
in

with (import (hostpkgs.fetchFromGitHub {
  owner = "NixOS";
  repo = "nixpkgs";
}) { });

mkShell {
  buildInputs = [
    (python36.withPackages (ps: with ps; [
      cvxpy
    ]))
  ];
}
```

With the nixos commit being this PR: https://github.com/NixOS/nixpkgs/pull/59518/commits/0115285de38b27ed9c3cf900ec3abff0a7a37632

I also copied your default.nix, placed it explicitly in the `withPackages` section and bumped the sha256+version to 1.0.25, but still got the same error.

**Comment by teh** [comment]
I did a rebase against master and several rebuilds / tests and still can't reproduce. One thing that's a bit confusing is that you got numerical errors first on cvxopt and now cvxpy. 

The only similar incident I can remember recently was related to an openblas bug. What's your CPU architecture (/proc/cpuinfo)? I'm on `Intel(R) Core(TM) i7-5557U CPU @ 3.10GHz`

**Comment by stites** [comment]
Quick followup -- I just built this on an `Intel i7-8665U (8) @ 4.800GHz` successfully, but got the above error when building on `Intel i9-7900X (20) @ 4.500GHz`. Both were built on NixOS, as well as x64 architectures.

Let me know if I can help debug this further. I'm not quite sure what to do next.

**Comment by drewrisinger** [comment]
I think this was superseded by #78898, which was just merged. Feel free to close @jonringer @FRidh @costrouc @teh

**Pull request closed without merging**